### ISSUE TITLE
[Bug] 自己训练PKINet-S模型性能差很多，复现官方模型的性能也差0.2个百分点

### ISSUE THREAD

**Issue by Terry-Zheng** [PKINet/PKINet#5]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmrotate/issues) and [Discussions](https://github.com/open-mmlab/mmrotate/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmrotate.readthedocs.io/en/1.x/notes/4_faq.html) but cannot get the expected help.
- [x] The bug has not been fixed in the [latest version (master)](https://github.com/open-mmlab/mmrotate) or [latest version (1.x)](https://github.com/open-mmlab/mmrotate/tree/dev-1.x).

### Task

I'm using the official example scripts/configs for the officially supported tasks/models/datasets.

### Branch

master branch https://github.com/open-mmlab/mmrotate

### Environment

> sys.platform: linux
> Python: 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:10) [GCC 10.3.0]
> CUDA available: True
> GPU 0,1: NVIDIA RTX A6000
> CUDA_HOME: /usr/local/cuda
> NVCC: Cuda compilation tools, release 11.8, V11.8.89
> GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
> PyTorch: 1.13.0a0+d0d6b1f
> PyTorch compiling details: PyTorch built with:
>   - GCC 9.4
>   - C++ Version: 201402
>   - Intel(R) Math Kernel Library Version 2020.0.4 Product Build 20200917 for Intel(R) 64 architecture applications
>   - Intel(R) MKL-DNN v2.6.0 (Git Hash N/A)
>   - OpenMP 201511 (a.k.a. OpenMP 4.5)
>   - LAPACK is enabled (usually provided by MKL)
>   - NNPACK is enabled
>   - CPU capability usage: NO AVX
>   - CUDA Runtime 11.8
>   - NVCC architecture flags: -gencode;arch=compute_52,code=sm_52;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_90,code=sm_90;-gencode;arch=compute_90,code=compute_90
>   - CuDNN 8.6
>   - Magma 2.6.2
>   - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.8, CUDNN_VERSION=8.6.0, CXX_COMPILER=/usr/bin/c++, CXX_FLAGS=-fno-gnu-unique -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX512=1, TORCH_VERSION=1.13.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=ON, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 
> 
> TorchVision: 0.14.0a0
> OpenCV: 4.5.5
> MMCV: 1.7.2
> MMCV Compiler: GCC 9.4
> MMCV CUDA Compiler: 11.8
> MMRotate: 0.3.4+022b30c

### Reproduces the problem - code sample

- 代码来源：
    - 仓库地址：[NUST-Machine-Intelligence-Laboratory/PKINet](https://github.com/NUST-Machine-Intelligence-Laboratory/PKINet)
    - 具体commitid：022b30c125529f7f9007cca52a3649e728bc8b4f（Mon Mar 11 09:42:06 2024 +0800）
    - 大致区别：落后本仓库大概4个commit，但是我看了基本没有区别
        - （基本上是针对保存模型的上限`checkpoint_config = dict(interval=1, max_keep_ckpts=3)`这一点进行改动）
- 自己的改动：
    - 代码的修改：基本上我能确定不会影响训练
        - 一个是work_dir下生成了个timestamp文件夹，这个不会影响到实际训练才对
        - 另一个是在`test.py`那里生成`dota.zip`的地方加了个参数
            ```python
            # 原tools/test.py约253行处强加了submission_dir
            if args.format_only:
                dataset.format_results(outputs, submission_dir=args.work_dir+'/dota', **kwargs)
            ```
    - 配置上的修改：
        - 不影响训练的应该是val_interval和max_keep_ckpts，我让它全部保存下来，且都验证一遍，以及数据集的地址
            - 这些理论上应该都是不会影响结果的
        - **唯一可能造成影响的最多是data这里，我调高了bs**
            > data = dict(
            >    samples_per_gpu=5,
            >    workers_per_gpu=20,...

### Reproduces the problem - command or script

```bash 
CUDA_VISIBLE_DEVICES=4 python -W ignore  tools/test.py configs/pkinet/pkinet-s_fpn_o-rcnn-dotav1-ss_le90.py work_dirs/pkinet-s_fpn_o-rcnn-dotav1-ss_le90/20240328_185012/best_mAP_epoch_29.pth --format-only
```

### Reproduces the problem - error message

我用这个test.py脚本生成的dota.zip上传到dota服务器后，得到的结果只有77.13%(ss)

> This is your evaluation result for task 1 (VOC metrics):
> 
> mAP: 0.7712878288276664
> ap of each class: 
> plane:0.8921556312426722,
> baseball-diamond:0.8541395188563535,
> bridge:0.552340458413297,
> ground-track-field:0.7472443176791004,
> small-vehicle:0.7488734665445337,
> large-vehicle:0.7933568002507358,
> ship:0.8803389235566501,
> tennis-court:0.9085144658620287,
> basketball-court:0.878562722403226,
> storage-tank:0.8681675686981466,
> soccer-ball-field:0.6131774958083259,
> roundabout:0.7122898908750352,
> harbor:0.776574859717641,
> swimming-pool:0.7373921983498644,
> helicopter:0.6061891141573851

用官方模型跑了一遍，也差了0.2个百分点

> This is your evaluation result for task 1 (VOC metrics):
> 
> mAP: 0.781947838570083
> ap of each class: plane:0.8971543686873233, 
> baseball-diamond:0.8423845433539111, 
> bridge:0.5582692134059756, 
> ground-track-field:0.7454337497397737, 
> small-vehicle:0.8024679416450173, 
> large-vehicle:0.8447171704945994, 
> ship:0.8812271740884957, 
> tennis-court:0.9088170028463273, 
> basketball-court:0.8757038432619434, 
> storage-tank:0.8609081933031836, 
> soccer-ball-field:0.6686228165256446, 
> roundabout:0.7029886474683681, 
> harbor:0.7747976536568887, 
> swimming-pool:0.7363188263990672, 
> helicopter:0.6294064336747264

### Additional information

1. 想问下你们在实际训练和验证时候的执行脚本是怎样的，相比你们发布的config版本是否有变化？我想随机种子之类的应该不会影响到1个多点才对
2. 我自己训出来的这版模型已经放到百度网盘(30天链接)：https://pan.baidu.com/s/13WLz_fRnp_MwqQ9r29Vpfw?pwd=yftq

**Comment by Terry-Zheng** [comment]
另外提供一组用官方的PKINet-S模型跑出来的本地trainval评估结果
- 执行脚本
    ```bash
    CUDA_VISIBLE_DEVICES=4 python -W ignore  tools/test.py configs/pkinet/pkinet-s_fpn_o-rcnn-dotav1-ss_le90.py work_dirs/pkinet_s_o-rcnn_dotav1-ss.pth --eval mAP
    ```
- 结果：
    | class              | gts   | dets  | recall | ap    |
    |:------------------:|:-----:|:-----:|:------:|:-----:|
    | plane              | 18788 | 19882 | 0.974  | 0.908 |
    | baseball-diamond   | 1087  | 1353  | 0.991  | 0.906 |
    | bridge             | 4181  | 6647  | 0.923  | 0.877 |
    | ground-track-field | 733   | 1264  | 0.982  | 0.905 |
    | small-vehicle      | 58868 | 73779 | 0.917  | 0.881 |
    | large-vehicle      | 43075 | 58607 | 0.974  | 0.904 |
    | ship               | 76153 | 82549 | 0.936  | 0.903 |
    | tennis-court       | 5923  | 6449  | 0.998  | 0.909 |
    | basketball-court   | 1180  | 1430  | 0.999  | 0.909 |
    | storage-tank       | 13670 | 12967 | 0.793  | 0.726 |
    | soccer-ball-field  | 827   | 1409  | 0.973  | 0.904 |
    | roundabout         | 973   | 1244  | 0.965  | 0.908 |
    | harbor             | 15468 | 20246 | 0.956  | 0.898 |
    | swimming-pool      | 3836  | 5137  | 0.966  | 0.897 |
    | helicopter         | 1189  | 1446  | 0.990  | 0.909 |
    | mAP                |       |       |        | 0.890 |

    {'mAP': 0.8896539211273193}

**Comment by PKINet** [comment]
I test the uploaded [PKINet-S](https://1drv.ms/u/c/9ce9a57f1a400a74/EXQKQBp_pekggJxsAAAAAAABWxHIx4vrnkZsRy1JW3BRaw?e=e07o7V) model and resubmit to the DOTA server, getting the same result in the paper:

> This is your evaluation result for task 1 (VOC metrics):
> mAP: 0.7839000653535957
> ap of each class: plane:0.8971739136636208, baseball-diamond:0.8420231218923928, bridge:0.5581433322673469, ground-track-field:0.7763072070435807, small-vehicle:0.8025395535863329, large-vehicle:0.8445838033051205, ship:0.881191617020791, tennis-court:0.908817056350211, basketball-court:0.8757115301592425, storage-tank:0.8607164857451293, soccer-ball-field:0.6686228165256446, roundabout:0.7023511357369185, harbor:0.7747472935220308, swimming-pool:0.7361656798108478, helicopter:0.6294064336747264
> COCO style result:
> AP50: 0.7839000653535957
> AP75: 0.5201910719020753
> mAP: 0.48281526580938666

Our actual training code is consistent with the [provided code](https://github.com/PKINet/PKINet/tree/main). For training, we use the following script:
`./tools/dist_train.sh ${CONFIG_FILE} ${GPU_NUM}`.
For testing, we use the following script：
`./tools/dist_test.sh ${CONFIG_FILE} ${CHECKPOINT_FILE} ${GPU_NUM}`.

We fixed the seed to 0. All models were trained only once on DOTA.

I have the following suggestions regarding the differences in the performance:

1. Make sure you follow the [data-splitting process](https://github.com/open-mmlab/mmrotate/blob/main/tools/data/dota/README.md).
2. Try the exact batchsize and learning rate. We use 4 Tesla V100-32G GPUs, and set the batchsize and learning rate to 2*4 and 0.0002 for training PKINet-S, respectively.
3. Set the seed to 0.
4. Submit the inference results of the final epoch.

**Comment by ZhenboZhao77** [comment]
你好请问能把训练的log文件发出来看一下吗？我也使用这个backbone 想看前几轮训练效果！谢谢

**Comment by PKINet** [comment]
> 你好请问能把训练的log文件发出来看一下吗？我也使用这个backbone 想看前几轮训练效果！谢谢

We have provided the log file in another issue #11.

This issue is temporarily closed due to inactivity. Feel free to reopen it.

**Comment by CrazyBrick** [comment]
@PKINet 你好，我在复现过程中发现训练得到的性能也比较差，请问有哪些需要关注的呢？bs没改，因为用的是3卡，所以调小了lr，其他的没有什么影响训练的改动
